### ISSUE TITLE
steps taking too long?

### ISSUE THREAD

**Issue by alphaatm** [tdrussell/diffusion-pipe#228]
im on a 4090 and 1 step each is taking around 300-600 seconds and i have 0 clue why. and my ram is at around 118gb out of 128gb am i missing something?  im training on flux dev. the links below should take u to the code just trying to make it cleaner.

flux.toml file: https://paste.pythondiscord.com/DOQA

dataset.toml file: https://paste.pythondiscord.com/CJ3Q

what im getting in cmd: https://paste.pythondiscord.com/XGZQ

**Comment by tdrussell** [comment]
Flux will not fit in VRAM on a single 4090 in bf16. Make sure to use ```transformer_dtype = 'float8'```. And actually, at 1024 resolution it might not fit even in fp8, depending on lora rank. But at rank 8 it might be fine. The reason it is working at all is probably because the Windows Nvidia driver has a feature where it starts auto-offloading stuff to RAM. This causes a massive slowdown and generally you don't want this enabled.

**Comment by alphaatm** [comment]
> Flux will not fit in VRAM on a single 4090 in bf16. Make sure to use `transformer_dtype = 'float8'`. And actually, at 1024 resolution it might not fit even in fp8, depending on lora rank. But at rank 8 it might be fine. The reason it is working at all is probably because the Windows Nvidia driver has a feature where it starts auto-offloading stuff to RAM. This causes a massive slowdown and generally you don't want this enabled.

i did what u said in my flux.toml https://paste.pythondiscord.com/UMVQ but now im getting this in my cmd https://paste.pythondiscord.com/6IVQ

**Comment by tdrussell** [comment]
You have to leave the dtype as bfloat16 and only change transformer_dtype.

**Comment by alphaatm** [comment]
> You have to leave the dtype as bfloat16 and only change transformer_dtype.

okay and would u be able to tell me how i would know which epoch is the best or how im able to do samples? so i can know what to pick?

**Comment by alphaatm** [comment]
> You have to leave the dtype as bfloat16 and only change transformer_dtype.

<img width="404" alt="Image" src="https://github.com/user-attachments/assets/406f6ae7-69a8-43a2-90c8-d00e55ed21a6" />

and also would u be able to tell me if i have someonthing wrong because i dont see any loras for the epochs in my out folder its just giving me this in ss

**Comment by tdrussell** [comment]
You will just have to test the various epochs yourself. You can't do sampling during training, and probably this training script will never support that for various reasons. But you can use a separate validation dataset and look at validation loss, if you want to go that route. The example config shows how to do this.

**Comment by alienzed** [comment]
> You will just have to test the various epochs yourself. You can't do sampling during training, and probably this training script will never support that for various reasons. But you can use a separate validation dataset and look at validation loss, if you want to go that route. The example config shows how to do this.

and I was going to request that you build in sampling during training... damn :P

@alphaatm You'll want to chose a range of epochs to test based on your loss curve. Generally speaking the sweet spots are going to be either right before, right at or right after a real plateau. A real plateau will consist of sideways loss lasting for ~10% of your run. Most people seem to stop training well before that and just release, which is why you'll see so many bad seeds and/or horrors, assuming the dataset and captions weren't mangled. For instance, if a single caption is bad, misrepresenting something in the scene, that's going to cause noise or worse even and hurt the LORA quite a bit, Triple or even quadruple check, manually, your captions.
I have training notes here: https://civitai.com/articles/13187/hunyuan-video-lora-training
I'll try to add a section on knowing when to stop since that's key and something I feel like I'm a lot better at now. Short answer though, you're probably stopping too early :P
Annoying but real answer: you have to test your the models manually to know for sure, since high loss doesn't necessarily imply bad results (i've had great LORAs saturate at 0.035) and low loss doesn't necessarily mean good results (overfitting or worse!)
